olp-cpp-sdk-dataservice-read/src/VersionedLayerClient.cpp

```diff
@@ -0,0 +1,202 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include <olp/dataservice/read/VersionedLayerClient.h>
+
+#include <olp/core/client/OlpClientFactory.h>
+#include <olp/core/thread/TaskScheduler.h>
+
+#include <olp/core/client/OlpClient.h>
+#include <olp/core/context/Context.h>
+
+#include "generated/api/BlobApi.h"
+#include "generated/api/QueryApi.h"
+
+#include "ApiClientLookup.h"
+#include "generated/model/Api.h"
+#include "olp/core/logging/Log.h"
+#include "olp/dataservice/read/Condition.h"
+#include "repositories/ExecuteOrSchedule.inl"
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+namespace {
+constexpr auto kLogTag = "VersionedLayerClient";
+}
+
+class VersionedLayerClient::Impl final {
+ public:
+  Impl(client::OlpClientSettings client_settings, client::HRN hrn)
+      : client_settings_(std::move(client_settings)), hrn_(std::move(hrn)) {}
+
+  ~Impl() = default;
+
+ private:
+  template <typename Result, typename Error>
+  static bool wait_and_check(
+      Condition& condition, const olp::client::CancellationContext& context,
+      const olp::client::ApiResponse<Result, Error>& response,
+      DataResponseCallback callback) {
+    if (!condition.Wait() && !context.IsCancelled()) {
+      callback({{olp::client::ErrorCode::RequestTimeout,
+                 "Network request timed out.", true}});
+      return false;
+    }
+    if (!response.IsSuccessful()) {
+      callback({{olp::client::ErrorCode::ServiceUnavailable,
+                 "Request unsuccessful.", true}});
+      return false;
+    }
+    return true;
+  }
+
+ public:
+  olp::client::CancellationToken GetDataByPartitionId(
+      const DataRequest& data_request, DataResponseCallback callback) {
+    auto context = std::make_shared<olp::client::CancellationContext>();
+    olp::client::CancellationToken token(
+        [context]() mutable { context->CancelOperation(); });
+
+    auto hrn = hrn_;
+    auto client_settings = client_settings_;
+    auto partition_id = data_request.GetPartitionId().value_or("");
+    auto layer_id = data_request.GetLayerId();
+    auto layer_version = data_request.GetVersion();
+    auto olp_client = olp::client::OlpClientFactory::Create(client_settings);
+
+    auto task = [context, callback, partition_id, olp_client, hrn, layer_id,
+                 layer_version]() {
+      Condition condition(*context);
+
+      // Step 1. Get query service
+      OLP_SDK_LOG_DEBUG_F(kLogTag, "step 1");
+
+      ApiClientLookup::ApiClientResponse apis_response;
```

Comment:
What do you mean?

Reply:
We cannot put everything inside the impl as it will explode. Currently, the functionality to get the data is spread over three classes: CatalogClientImpl, DataRepository, and DataCacheRepository. If you put everything in one class it will become a nightmare to manage.